internal/pkg/agent/application/upgrade/manual_rollback.go

```diff
@@ -0,0 +1,258 @@
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// rollback
+	watcherCmd, err := u.watcherHelper.InvokeWatcher(u.log, watcherExecutable, "--rollback", versionedHomeToRollbackTo)
+	if err != nil {
+		if !updateMarkerExistsBeforeRollback {
+			// best effort: cleanup the fake update marker
+			cleanupErr := os.Remove(updateMarkerPath)
+			if cleanupErr != nil {
+				u.log.Errorf("Error cleaning up fake upgrade marker: %v", cleanupErr)
+			}
+		} else {
+			// attempt to resume the  "normal" watcher to continue watching
+			_, restoreWatcherErr := u.watcherHelper.InvokeWatcher(u.log, watcherExecutable)
+			if restoreWatcherErr != nil {
+				u.log.Errorf("Error resuming watch after rollback error : %v", restoreWatcherErr)
+			}
+		}
+
+		return nil, fmt.Errorf("starting rollback command: %w", err)
+	}
+
+	u.log.Infof("rollback command started successfully, PID: %d", watcherCmd.Process.Pid)
+	return nil, nil
+}
+
+func rollbackUsingAgentInstalls(log *logger.Logger, watcherHelper WatcherHelper, source availableRollbacksSource, topDir string, now time.Time, rollbackVersion string, markUpgrade markUpgradeFunc) (string, string, error) {
+	// read the available installs
+	availableRollbacks, err := source.Get()
+	if err != nil {
+		return "", "", fmt.Errorf("retrieving available rollbacks: %w", err)
+	}
+	// check for the version we want to rollback to
+	var targetInstall string
+	var targetTTLMarker TTLMarker
+	for versionedHome, ttlMarker := range availableRollbacks {
+		if ttlMarker.Version == rollbackVersion && now.Before(ttlMarker.ValidUntil) {
+			// found a valid target
+			targetInstall = versionedHome
+			targetTTLMarker = ttlMarker
+			break
+		}
+	}
+
+	if targetInstall == "" {
+		return "", "", fmt.Errorf("version %q not listed among the available rollbacks: %w", rollbackVersion, ErrNoRollbacksAvailable)
+	}
+
+	prevAgentParsedVersion, err := version.ParseVersion(targetTTLMarker.Version)
+	if err != nil {
+		return "", "", fmt.Errorf("parsing version of target install %+v: %w", targetInstall, err)
+	}
+
+	// write out a fake upgrade marker to make the upgrade details state happy
+	currentHome := paths.VersionedHome(topDir)
+	relCurVersionedHome, err := filepath.Rel(topDir, currentHome)
+	if err != nil {
+		return "", "", fmt.Errorf("getting current install home path %q relative to top %q: %w", currentHome, topDir, err)
+	}
+	curAgentInstall := agentInstall{
+		parsedVersion: agtversion.GetParsedAgentPackageVersion(),
+		version:       release.VersionWithSnapshot(),
+		hash:          release.Commit(),
+		versionedHome: relCurVersionedHome,
+	}
+
+	prevAgentInstall := agentInstall{
+		parsedVersion: prevAgentParsedVersion,
+		version:       targetTTLMarker.Version,
+		hash:          targetTTLMarker.Hash,
+		versionedHome: targetInstall,
+	}
+
+	upgradeDetails := details.NewDetails(release.VersionWithSnapshot(), details.StateRequested, "" /*action.ID*/)
```

Comment:
I'm wondering if we should introduce a new state, something like `UPG_REQUESTED_ROLLBACK` to distinguish it from `UPG_REQUESTED`?  Might make it easier to reason about the state of the upgrade when looking at a diagnostic.  That also makes the upgrade marker created here representative of the state of the upgrade and not just a fake one.

Reply:
Give the current state of upgrade details I am hesitant to add another state there. This set here is just to accomodate the initialization code we have in the startup of the rolled back agent.
The `UPG_REQUESTED` is the same state that agent will **unconditionally** enter upon receiving an upgrade request/action of any kind